### PR TITLE
CISCO-NTP-MIB

### PR DESCRIPTION
**Pull request by adaniels21487**
A new Discovery/Poller module to collect NTP statistics from devices which support the CISCO-NTP-MIB

- Discovered peers are stored using components
- Statistics are displayed using 'applications'
- A critical alarm is raised when a stratum of 16 is reported.

### CODE REVIEW

**Comment by laf** [comment]
All looks good except that I think the disco and poller modules should just be called ntp as cisco based devices won't be the only thing that supports them.

We are also just about to merge a PR which tidies up the rrdtool functions in polling/discovery so you will need to update the code when that's merged.

**Comment by adaniels21487** [comment]
Hi @laf 
Re: rrdtool function.
No worries. I will update when its merged.

Re: rename the modules.
The MIB is Cisco proprietry and part of the Cisco management tree (1.3.6.1.4.1.9.9.168), I dont think anything non-Cisco will be implementing this MIB but happy to rename if you like.
Please confirm.

**Comment by laf** [comment]
Sorry I didn't mean that other vendors will use this mib, more that other vendors will have an implementation of NTP so we should make the module re-usable.

Like for instance the bgp polling, that does cisco and juniper with an if/elseif. So in the disco / poller module for this have ntp.inc.php, that then just does

```php
if ($device['os'] == "cisco") {
    require_once 'includes/discovery/ntp/cisco.inc.php';
}

```

**Comment by laf** [comment]
Ok, big PR merged. If you can update your code to match that would be ace.

**Comment by adaniels21487** [comment]
Hi @laf 
Understood, will do.

**Comment by adaniels21487** [comment]
Hi @laf 
Please let me know if this is what you had in mind?

**Comment by laf** [comment]
Initial look is what I was trying to say yes :)

**Comment by adaniels21487** [comment]
I dont believe this is still a blocker.

File: html/includes/graphs/device/ntp_delay.inc.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        $rrd_additions .= " DEF:DS" . $count . "=" . $rrd_filename . ":delay:AVERAGE ";
+        $rrd_additions .= " LINE1.25:DS" . $count . "#" . $color . ":'" . str_pad(substr($array['peer'].' (s)',0,15),15) . "'" . $stack;
+        $rrd_additions .= " GPRINT:DS" . $count . ":LAST:%7.0lf ";
+        $rrd_additions .= " GPRINT:DS" . $count .    ":MIN:%7.0lf ";
+        $rrd_additions .= " GPRINT:DS" . $count . ":MAX:%7.0lf\\\l ";
+        $count++;
+    }
+}
+
+if ($rrd_additions == "") {
```
Comment (laf) [inline html/includes/graphs/device/ntp_delay.inc.php]: Seems odd to check and not do anything, may as well just !empty() or something and then append to $rrd_options.

Not a show stopper but would be nice to change it post PR.

File: html/includes/graphs/device/ntp_dispersion.inc.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        $rrd_additions .= " DEF:DS" . $count . "=" . $rrd_filename . ":dispersion:AVERAGE ";
+        $rrd_additions .= " LINE1.25:DS" . $count . "#" . $color . ":'" . str_pad(substr($array['peer'].' (s)',0,15),15) . "'" . $stack;
+        $rrd_additions .= " GPRINT:DS" . $count . ":LAST:%7.0lf ";
+        $rrd_additions .= " GPRINT:DS" . $count .    ":MIN:%7.0lf ";
+        $rrd_additions .= " GPRINT:DS" . $count . ":MAX:%7.0lf\\\l ";
+        $count++;
+    }
+}
+
+if ($rrd_additions == "") {
```
Comment (laf) [inline html/includes/graphs/device/ntp_dispersion.inc.php]: Same here.

**Comment by laf** [comment]
I'm merging this but we really really need to have all tables paginated and I was considering not merging this at present.

However, can you please submit a new PR to resolve this.

**Comment by adaniels21487** [comment]
Thanks @laf 
Are you referring to:
html/pages/apps/ntp.inc.php
html/pages/device/apps/ntp.inc.php

Let me know and I'll get it done.

**Pull request merged**